### PR TITLE
Dynamically look for battery info under `/sys/class/power_supply` directory 

### PR DESCRIPTION
**Pull request by akshaykhoje**
## The issue #127 is hereby fixed as can be seen in the attached screenshot.

image

It is also made dynamic for the folder in `/sys/class/power_supply`. The code basically checks for any folder inside this one that starts with 'BA'. The required files of `status` and `capacity` are generally located in BAT0 or BAT1 folder here. The other case maybe of BAPM.

Thus the battery folder will always begin with 'BA' substring. The other folder will be for the adapter which doesn't start with 'BA' at all!

Hope this helps.

Revert back if any changes are required.

Thank you :)

### CODE REVIEW

File: src/fetch.cpp [hunk 1]
```diff
@@ -521,11 +530,25 @@ void print_bar(int battery)
  */
 void print_battery(string path)
 {
+    string dir_path = path;
+    string capacity_path = path;
+
+    for (const auto& entry : filesystem::directory_iterator(dir_path)) {
```
Comment (TanmayPatil105) [inline src/fetch.cpp:536]: Style changes: move curly braces below
it should be
```
for(; i; i++)
{

}
```

File: src/fetch.cpp [hunk 2]
```diff
@@ -521,11 +530,25 @@ void print_bar(int battery)
...
+        if (batteryCheck(dir_path)) {
+            string filename = entry.path().filename().string();
+            if (filename.substr(0, 2) == "BA") {
```
Comment (TanmayPatil105) [inline src/fetch.cpp:539]: style check

**Review by TanmayPatil105 — COMMENTED** [review]
Please do the following suggested changes.
Thanks @akshaykhoje!!

File: src/fetch.cpp [hunk 3]
```diff
@@ -480,13 +488,14 @@ bool isCharging(string path)
 /**
  * @brief Utility to print battery perecentage bar
  */
-void print_bar(int battery)
+void print_bar(string path, int battery)
```
[the remark this answers is not in the document]
Comment (akshaykhoje) [reply src/fetch.cpp:491]: Sure!

File: src/fetch.cpp [hunk 4]
```diff
@@ -480,13 +488,14 @@ bool isCharging(string path)
...
 {
     auto red = Crayon{}.bright().red();
     auto green = Crayon{}.bright().green();
 
     string emoji = "\n🔋 ";
-    if (isCharging("/sys/class/power_supply/BAT0/status"))
+    string status_path = path + "/status";
```
[the remark this answers is not in the document]
Comment (akshaykhoje) [reply src/fetch.cpp:497]: yeah

File: src/fetch.cpp (hunk 1 above)
Comment (akshaykhoje) [reply src/fetch.cpp:536]: Noted!

File: src/fetch.cpp (hunk 2 above)
Comment (TanmayPatil105) [reply src/fetch.cpp:539]: Move the curly braces below! 
See https://github.com/TanmayPatil105/procfetch/blob/main/src/fetch.cpp#L369

Comment (akshaykhoje) [reply src/fetch.cpp:539]: Gotcha!

Comment (TanmayPatil105) [reply src/fetch.cpp:539]: Yeah, you should add a iterator method to class `Path` that would return contents of that directory!
Maybe a `vector<Path>`. Then we iterate through the contents `vector` and check if "BA" directory is present

**Comment by akshaykhoje** [comment]
@TanmayPatil105 all the aforementioned changes have been made. Please review this PR.

File: src/fetch.cpp (hunk 2 above)
Comment (TanmayPatil105) [reply src/fetch.cpp]: Hey @akshaykhoje!
You missed this one

Comment (akshaykhoje) [reply src/fetch.cpp]: > Move the curly braces below! See https://github.com/TanmayPatil105/procfetch/blob/main/src/fetch.cpp#L369

image

This good?

Comment (akshaykhoje) [reply src/fetch.cpp]: > Yeah, you should add a iterator method to class `Path` that would return contents of that directory! Maybe a `vector<Path>`. Then we iterate through the contents `vector` and check if "BA*" directory is present

Okay will do this.

Comment (TanmayPatil105) [reply src/fetch.cpp]: Yeah, I was talking about adding a getDirectoryContens method  to class `Path` and then we iterate over the vector
something like

```cpp
class Path
{
   public:
     vector<Path> getDirectoryContents()
      {

      }
}
```

Comment (TanmayPatil105) [reply src/fetch.cpp]: Also, can you add unit tests for that method?
See https://github.com/TanmayPatil105/procfetch/blob/main/src/util.cpp#L66

Comment (akshaykhoje) [reply src/fetch.cpp]: > Yeah, I was talking about adding a getDirectoryContens method to class `Path` and then we iterate over the vector something like
> 
> ```c++
> class Path
> {
>    public:
>      vector<Path> getDirectoryContents()
>       {
> 
>       }
> }
> ```

Sure will try!

Comment (akshaykhoje) [reply src/fetch.cpp]: > Also, can you add unit tests for that method? See https://github.com/TanmayPatil105/procfetch/blob/main/src/util.cpp#L66

Okay lemme try.

**Comment by youpong** [comment]
The following will automatically format the source code, so try it out.
```
$ make -C src format
```

**Comment by akshaykhoje** [comment]
> The following will automatically format the source code, so try it out.
> 
> ```
> $ make -C src format
> ```

doesn't seem to work with me. Am I doing anything wrong here?
image

File: src/fetch.h [hunk 5]
```diff
@@ -295,6 +295,19 @@ class Path
     {
         return path.string();
     }
+
+    /**
+     * @returns a vector containing absolute paths to contents at the given path
+     */
+    vector<filesystem::path> getDirectoryContents()
```
Comment (TanmayPatil105) [inline src/fetch.h:302]: I think we should return `vector<Path>` and not `vector<filesystem::path>`

File: src/util.cpp [hunk 6]
```diff
@@ -88,6 +88,11 @@ static void test_Path()
     expect(false, p.isRegularFile(), "test -f "s + p.toString());
     expect(false, p.isExecutable(), "test -x "s + p.toString());
     expect(false, p.isDirectory(), "test -d "s + p.toString());
+
+    // directory is not empty
+    p = Path::of("/bin"s);
+    vector<filesystem::path> directoryContents = p.getDirectoryContents();
+    expect(true, !directoryContents.empty(), "Directory is not empty");
```
Comment (TanmayPatil105) [inline src/util.cpp:95]: can you write more tests to it?
- you can pass a non-existent directory as a parameter
- pass in a file name eg. `/bin/ls`
- ...

**Comment by TanmayPatil105** [comment]
Hey @akshaykhoje!
Can you pull the changes and see if it works?

**Comment by akshaykhoje** [comment]
Heyy @TanmayPatil105
The changes made seem to work now.

Attaching a screenshot for the same.

image

**Comment by akshaykhoje** [comment]
Just to add some notes for future references, I am adding a quick check used to bring about the effective change.

In the `printBattery()` function in fetch.cpp, the following block of code has been added which searches for the substring "BA" instead of "BAT1" or "BAT0". It is so because while researching, I came across a few other folders in the `/sys/class/power_supply/` folder, whose common initial string was still "BA", but the third letter differed.

Attaching the code image hereby.

image

**Pull request merged**